presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/RuleStore.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.SetMultimap;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.Set;
+
+public class RuleStore
+{
+    private final ListMultimap<Class<? extends PlanNode>, Rule> rules;
+
+    public RuleStore(ListMultimap<Class<? extends PlanNode>, Rule> rules)
+    {
+        this.rules = rules;
+    }
+
+    public List<Rule> getPossiblyMatchingRules(PlanNode planNode)
+    {
+        return ImmutableList.<Rule>builder()
+                .addAll(rules.get(planNode.getClass()))
+                .addAll(rules.get(null))
+                .build();
+    }
+
+    public static Builder builder()
+    {
+        return new Builder();
+    }
+
+    public static class Builder
+    {
+        private final SetMultimap<Class<? extends PlanNode>, Rule> rules = HashMultimap.create();
+
+        public Builder register(Set<Rule> newRules)
+        {
+            newRules.forEach(this::register);
+            return this;
+        }
+
+        public Builder register(Rule newRule)
+        {
+            rules.put(getPatternNodeClass(newRule), newRule);
+            return this;
+        }
+
+        @Nullable
+        private Class<? extends PlanNode> getPatternNodeClass(Rule rule)
+        {
+            Pattern pattern = rule.getPattern();
+            if (pattern instanceof Pattern.AnyNode) {
+                return null;
+            }
+            else if (pattern instanceof Pattern.ClassNodeBased) {
+                return ((Pattern.ClassNodeBased) pattern).getNodeClass();
+            }
+            throw new IllegalArgumentException("Unexpected Pattern: " + pattern);
```

Comment:
So `Pattern` being an interface one can implement is... a lie? Why don't you make it a class with a private/package-private ctor then?